### DISCUSSION TITLE
Opportunistic refactoring

### DISCUSSION THREAD

**Discussion by deltamarnix** [pyphoenix/pyphoenix-project#37] [Ideas]
We will need to make quite some adjustments in the FloPy code base. We will need to refactor the current code base and split out certain functionality into a more modular approach, before we can actually implement some of the new ideas that we have. This discussion aims to make a breakdown of the work that is still to be done. From there we can start estimating the work.

# Work with minimal impact on the API

## Refactor `PackageContainer` to composition over inheritance ✅ 

I have been playing with the idea of doing a refactor and keeping as much in place as possible. I agree that `PackageContainer` and `PackageInterface` are daunting classes, but we can separate the functionality in a more modular fashion to simplify our intentions.

What we see in the current flopy design, is the overuse of inheritance where composition would most likely be sufficient. We see that both simulation, model, and package are inheriting from `PackageContainer`, while having a `PackageContainer` as a field would be sufficient. Inheritance should be used when there is also functionality that expects any kind of `PackageContainer`, no matter what the subclass is. I have not found such code. In this case inheritance was used to share functionality among different types of classes. Which can also be achieved with composition, as the `PackageContainer` doesn't need anything from its subclass and there are no abstract methods in `PackageContainer`.

During https://github.com/modflowpy/flopy/pull/2324, I came to the point that it was very well possible to do this refactor, but I have doubts on which functions can actually be removed, because they are supposed to be for internal use only. There are options to either just remove them, keep them, or `@deprecate` them.

![image](https://github.com/user-attachments/assets/c6051ae2-2e62-471a-9d9b-816b33fdb7d7)

## Move functions out of `PackageInterface` that can be done modular

We could also look at other types of functions that could move out, like the `export` functionality. It would make more sense to do this in a similar fashion to `cattrs` with a module and single-dispatch functions. The functions that can go out of `PackageInterface` are: `plot()`, `export()`, and `check()`.

It is possible to do this with minimal user impact by keeping the `plot()` function, but `@deprecate` them in favor of the modular version. The `plot()` function in `PackageInterface` would simply point to the modular version.

## Add type hints to critical public classes

I do expect that refactoring is a tough job at the moment, because there are no type hints in the code. VS Code cannot follow the methods and properties when renaming or moving and doesn't give any error messages when functions no longer exist. I recommend adding type hints in the classes we want to refactor, and let mypy run over it.

This can be done with minimal impact on backwards compatibility, as we are just adding type hints, and not removing anything for the user.

## Move static methods to modules

Some functions are now decorated as `@staticmethod`, but they don't really apply to their class and could just as well be a global function in a module. E.g. `Package.add_to_dtype()` doesn't seem to use anything from the `Package` class and could just as well be a separate function.

Similar to making functions like `plot()` more modular, the same accounts for these `@staticmethod`s. We could keep the old ones for now and `@deprecate` them. Then simply point to the new modular version.

## Let generate_classes.py use `jinja` ✅ 

The code that is now generated with print statements and `write()` functions. It would be so much easier to use jinja, so it is clearer to read the initial file that is being used for generation. The jinja package handles template generation better than `write()` with raw strings.

## Generate dfn classes on user install

See #35.

This idea would have minimal impact on the user, as it is being performed during get-modflow. So no changes in the API.

# Work with likely more impact on the user

## Use XArray / Dask to speed up performance

See #49. 

## Use `attrs` for property generation

There is quite some boilerplate code for variable getters and setters. The `attrs` package could help with this and simplify the code.

Introducing attr itself and simply replacing the things that we have would not have any impact on the user's code, but if we decide to move / remove certain variables, it will.

**Comment by wpbonelli** [comment]
a few more:

- consolidate input specifications &mdash; i.e. is everything in `mfstructure.py` necessary? is it necessary to reproduce the DFN spec verbatim as input component classes currently do? and similar questions
- investigate `discretization` module... can it be cleanly "unplugged" from existing core and plugged into a new framework?
- find the line between internal and user-facing APIs (this may not be well-defined at the moment) and begin to separate these from one another where appropriate/helpful
- investigate error handling approaches... why are flopy errors currently so uninformative? what can be done in the near-term to improve them? this will help refactoring efforts
- investigate the model check mechanism... correctness? extensibility?

**Reply by deltamarnix** [reply]
The model check mechanism has a weird setup and can definitely be changed. There is now some kind of inheritance going on for mf6, and there are some very specific edge cases that require different `check` functions.

**Reply by wpbonelli** [reply]
after some discussions I think we can outline a plan for checks:

- deprecate current check mechanism now (3.9) and unhook it in a later minor release (3.10?)
- tentatively leave the check API in place, just warn it will become non-functional for the rest of 3.x until we..
- reimplement checks for flopy 4

**Reply by wpbonelli** [reply]
the existing grid modules are stable and probably outside the scope of this work, but we should devote some attention to synchronizing them with dis and other packages

**Reply by wpbonelli** [reply]
maybe we should consider deprecating `is_valid()` methods like we have done for the check mechanism... what is the difference between these?

**Reply by deltamarnix** [reply]
Coming back to MFStructure. Our internal usages mainly look at `sim_struct`, `flopy_dict`, and `model_types`. It seems like we could generate this information during code generation because it is static information, and remove MFStructure in its entirety.

**Reply by wpbonelli** [reply]
> It seems like we could generate this information during code generation because it is static information, and remove MFStructure in its entirety.

Excellent!

**Comment by wpbonelli** [comment]
Another: we have long discussed converting DFNs to TOML, YAML or some other standard data interchange format. Maybe also opportunities to shrink/revise the DFN specification

**Reply by wpbonelli** [reply]
some tentative steps here (summarizing from discussion)

1. DFN -> TOML conversion capability &check;

Ability to load DFNs and emit TOML. Also to structure the input specification, moving from flat to nested representation. Where currently a DFN file is a list of variables each of which can be part of another variable, an input definition is really a tree with composite variables (lists, records, unions) as internal nodes and scalars/arrays as leaves. This is the internal structure used for code generation and runtime introspection. Moving to structured definition files makes it unnecessary to inflate composite types from a flat list in Python. Nested TOML files still appear flat so they will look quite familiar.

2. Temporary/transparent TOML conversion in FloPy (in progress)

We discussed using FloPy as a staging ground for the new TOML format. The code generation mechanism can temporarily add an internal step to convert DFNs to TOML before loading them. This gives us a place to kick the tires until we are comfortable taking it elsewhere

3. Permanent migration to TOML in MF6

Migrate the DFNs to TOML in the MF6 repo. This will take a bit of work as DFNs are parsed to generate docs and source code, but the structured format will hopefully have similar benefits here as in FloPy

**Reply by wpbonelli** [reply]
we should version the file format too: v1 is DFN, v2+ as TOML

**Reply by wpbonelli** [reply]
a short-term action item: check which DFN variable attributes are used by flopy, and which by mf6... are any redundant or unnecessary?

**Reply by wpbonelli** [reply]
once we have trialed the new toml definition file format in flopy for some time, we feel it has stabilized, and we are ready to adopt it back in mf6, we could consider moving the toml conversion script and any dfn-related tooling to modflow-devtools so both flopy and mf6 can use it

**Reply by wpbonelli** [reply]
separate discussion in #47

**Comment by wpbonelli** [comment]
Another from discussion with @langevin-usgs and @jdhughes-usgs: review input file path management. Any component, many blocks, variables, and even parts of variables (e.g. array layers) can be associated with input files (see MF6IO appendix for more details on `open/close`)

**Reply by deltamarnix** [reply]
Design-wise, file path management seems pretty straight forward. A simulation contains simulation data, and the simulation data contains a file management object. During load, write, and run, this `MFFileMgmt` is used to get/set paths.

```mermaid
classDiagram
class MFSimulationBase {
  simulation_data: MFSimulationData
  set_sim_path(path:StrPath)
}
class MFSimulationData {
  mfpath: MFFileMgmt
}
class MFFileMgmt {
  simulation: MFSimulationBase
  existing_file_dict: dict[MFFilePath]
  + copy_files()
  + get_model_path()
  + get_sim_path()
  ~ get_updated_path()
  ~ strip_model_relative_path()
  ~ set_sim_path(path:StrPath)
  ~ unique_file_name(file_name, lookup)$
  ~ set_last_accessed_path()
  ~ set_last_accessed_model_path()
  ~ resolve_path()
}
note for MFFileMgmt "simulation reference can go, because there\nis a deprecation warning in set_sim_path\nthat it should become an internal function.\nThat would remove the circular dependency.\nIs copy_files really intended for public use?"

MFSimulationBase *--> "*" MFModel
MFSimulationBase *--> "1" MFSimulationData
MFSimulationData *--> "1" MFFileMgmt
MFSimulationBase <-- MFFileMgmt : calls set_sim_path()
MFModel ..> MFFileMgmt: calls static unique_file_name()
MFFileMgmt *--> "*" MFFilePath
```

Looking at this drawing, I think it would be best to move MFFileMgmt to an internal file (_file_manager.py), showing the intention directly to the user. Anything that would need to be made public for users should be a function in MFSimulationBase instead. Candidates are `copy_files()`, `get_model_path()`, and `get_sim_path()`.

**Reply by wpbonelli** [reply]
separate discussion in #53

**Comment by deltamarnix** [comment]
I was playing a bit with "Move static methods to modules". I figured that the `PackageContainer` now has two functions. One  for containing a list of all types of models, and one for containing a list of all types of packages. When we split this up, we see that `modflow_models` is actually never used and can go. And then we are only left with a meager dict `models_by_type`. I would move that to a very simple module like:

```py
# _model_factory.py with an underscore so it's internal

_models_by_type: dict[str, type["MFModel"]] = {}

def register_model_type(model_type: str, model_class: type["MFModel"]) -> None:
    _models_by_type[model_type] = model_class

def get_model_class(model_type: str) -> type["MFModel"] | None:
    _models_by_type.get(model_type)
```

I don't think it needs to be anything more, and it can be internal this way. We can do a similar setup for the parts of package, because that does exactly the same.

**Comment by wpbonelli** [comment]
we can do an incremental deployment once we have a full initial prototype: first internally (flopy tests, mf6 tests, mf6 example models) before considering a public release

**Reply by wpbonelli** [reply]
separate discussion in #68

**Comment by wpbonelli** [comment]
eventually we want to rewrite tutorials/examples where each is more focused/finer-grained/shorter

**Reply by wpbonelli** [reply]
[mf6 tutorials](https://flopy.readthedocs.io/en/stable/tutorials.html#modflow-6) maybe a good place to start, topics including

- [ ] components (simulations, models, packages, subpackages)
- [ ] variables (describe all types, then demo scalars)
- [ ] composite variables (records and unions)
- [ ] list/table variables
- [ ] array variables
- [ ] transient data
- [ ] observations
- [ ] time(array)series
- [ ] input file management
- [ ] io format (binary/ascii, etc)
- [ ] io performance optimization

linked PRs

- #45
- #46

**Comment by deltamarnix** [comment]
I was looking a bit at `VerbosityLevel` and figured that there is a lot of custom logging code. We should use the default python logger instead and get rid of `VerbosityLevel`, `flopy.export.netcdf.Logger`, and more. There are now a bunch of `print()` statements that should be changed to a `logging.Logger`.

**Reply by wpbonelli** [reply]
separate discussion in #51

**Comment by wpbonelli** [comment]
Some items on the topic of parity between flopy3 and flopy4 input files... in general we want to shoot for functionally identical, with possible cosmetic differences.

Floating point precision:
- determine what existing conventions flopy3 has for floating point precision in input files (and if not adequately documented, do so)
- keeping in mind requirement `FUNC-14`, decide on flopy4 floating point precision output conventions 

Other possible criteria to consider:
- comments
- casing
- parameter order
- whitespace

**Reply by wpbonelli** [reply]
related: whether to support wrapped arrays.. is this needed, since mf6 no longer has a line length restriction?

**Reply by wpbonelli** [reply]
related: currently flopy allows comments at the end of lines, with no comment character &mdash; consider requiring end of line comments to start with e.g. "#"?

**Reply by wpbonelli** [reply]
separate discussion in #69
